### PR TITLE
Gps tests 1

### PR DESCRIPTION
**Pull request by sunny-goel**
### ✍🏽Description:

- [x] **Add GPS Start/Stop Services Test cases**
- [x] Modify **build.gradle** file by Adding required dependencies and Test Runner to setup creation & execution of tests 
- [x] **Add MockLocationProvider**
- [x] **Setup mockWebServer** for upcoming PING service tests 
- [x] **Add Utility classes** like BaseTestCase.java , AssertUtils , LogUiTest , TestConstants , UiUtils .. 
- [x] Add new annotation **AutoTest_StartStopServices.java**
- [x] **Added a debug check in MainActivity** to initiate services . 
- [x] **Updated gradle version**
- [x] **Added New Build Type (QA) for devs**

 
### 🆕 Tests Built:

image

### 📸Flakiness Test Runs

image

 
### Links:
[Clubhouse ticket](https://app.clubhouse.io/shuttl/story/7637/ca-build-tests-for-new-gps-sdk)

### CODE REVIEW

File: app/src/androidTest/java/mockLocationUtils/MockLocationProvider.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
+package mockLocationUtils;
```
Comment (nikhil-roy) [inline app/src/androidTest/java/mockLocationUtils/MockLocationProvider.java:1]: try using kotlin. it reduces the everyday bugs

File: app/src/main/java/com/shuttl/packagetest/MainActivity.kt [hunk 2]
```diff
@@ -52,12 +52,16 @@ class MainActivity : AppCompatActivity() {
         setContentView(R.layout.activity_main)
         requestLocationPermission()
 
-        val intent = Intent(this, LocationPingService::class.java)
-        intent.action = "STOP"
+        if (!BuildConfig.BUILD_TYPE.equals("debug"))
```
Comment (nikhil-roy) [inline app/src/main/java/com/shuttl/packagetest/MainActivity.kt:55]: which flavour you will be using?

File: locationsyncservice/build.gradle [hunk 3]
```diff
@@ -17,6 +17,25 @@ android {
         testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
     }
 
+    buildTypes {
```
Comment (nikhil-roy) [inline locationsyncservice/build.gradle:20]: do we need this in the module?

File: app/src/main/java/com/shuttl/packagetest/MainActivity.kt (hunk 2 above)
Comment (sunny-goel) [reply app/src/main/java/com/shuttl/packagetest/MainActivity.kt:55]: We are using Debug . 
I have created one more Build Type **QA** .

File: locationsyncservice/build.gradle (hunk 3 above)
Comment (sunny-goel) [reply locationsyncservice/build.gradle:20]: When I am building TestApplication module with QA BuildType , I was getting this error

image

Therefore, Have to add similar build Type in Other module as well .

File: .idea/misc.xml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_7" project-jdk-name="JDK" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" project-jdk-name="JDK" project-jdk-type="JavaSDK">
```
Comment (Saransh-Sharma) [inline .idea/misc.xml]: You should not be pushing this file. This is auto generated by the IDE & could be different for every dev. revert changes & add .idea/ folder to .gitignore.

File: app/build.gradle [hunk 5]
```diff
@@ -7,20 +7,40 @@ apply plugin: 'kotlin-android-extensions'
...
+        }
+        testInstrumentationRunnerArguments clearPackageData: 'true'
     }
     buildTypes {
         release {
             minifyEnabled false
             proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
         }
+
+        qa {
```
Comment (Saransh-Sharma) [inline app/build.gradle]: Let's rename this for dev as its for devs primarily.

File: app/src/androidTest/java/saveServiceTests/StartStopService.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+
+    @AutoTest_StartStopServices
+    @Test
+    public void verifyStartStopLocationSaveService() {
+
+        // Start Location Save Service
+        LogUITest.debug("Starting 'Save Location Service'");
+        LocationsHelper.INSTANCE.startLocationSaveService(activityTestRule.getActivity().getApplication(), locationConfigs);
```
Comment (Saransh-Sharma) [inline app/src/androidTest/java/saveServiceTests/StartStopService.java:113]: Why do we not use the init() method which the apps use to start GPS SDK for all these start-stop tests?

File: app/src/androidTest/java/testUtils/TestConstants.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package testUtils;
+
+import testUtils.mockWebServer.MockWebUtils;
+
+public class TestConstants
+{
+
+    public static String VEHICLE_NUMBER = "";
```
Comment (Saransh-Sharma) [inline app/src/androidTest/java/testUtils/TestConstants.java:8]: Does the GPS SDK know what a vehicle number is?

**Review by Saransh-Sharma — CHANGES_REQUESTED** [review]
Another great PR ! Nice work!

File: app/src/androidTest/java/saveServiceTests/StartStopService.java (hunk 6 above)
Comment (sunny-goel) [reply app/src/androidTest/java/saveServiceTests/StartStopService.java:113]: So there are 3 different tests . 
image

Kindly read the description of very first test. 
First test is all about starting services from init module (which other apps call to start the services)

File: .idea/misc.xml (hunk 4 above)
Comment (sunny-goel) [reply .idea/misc.xml]: Done

File: app/build.gradle (hunk 5 above)
Comment (sunny-goel) [reply app/build.gradle]: Done

File: app/src/androidTest/java/testUtils/TestConstants.java (hunk 7 above)
Comment (sunny-goel) [reply app/src/androidTest/java/testUtils/TestConstants.java:8]: Yes, Please refer the screenshot of How consumer app starts services using init module . 
image

It is just a value which is to be stored in database. There are no logical checks on this field.

File: .gitignore [hunk 8]
```diff
@@ -83,3 +83,6 @@ lint/generated/
 lint/outputs/
 lint/tmp/
 # lint/reports/
+
+#idea files
+.idea/misc.xml
```
Comment (Saransh-Sharma) [inline .gitignore]: also add the entire .idea folder

Like `.idea/`

File: app/src/androidTest/java/mockLocationUtils/MockLocationProvider.java (hunk 1 above)
Comment (Saransh-Sharma) [reply app/src/androidTest/java/mockLocationUtils/MockLocationProvider.java:1]: We'll port mock location class to Kotlin once have some tests for GPS SDK.

File: app/src/androidTest/java/saveServiceTests/StartStopService.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+    @AutoTest_StartStopServices
+    @Test
+    public void verifyStartStopLocationServicesViaInitModule() {
+
+        Intent intent = new Intent(appContext, LocationPingService.class);
+        intent.setAction("STOP");
+        // Start Both Services Via Init module .
+        LogUITest.debug("Starting 'Save Location Service' & 'Ping Location Service Via Init Module'");
+        LocationsHelper.INSTANCE.initLocationsModule(activityTestRule.getActivity().getApplication(), null, locationConfigs, locationPingServiceCallback, intent);
```
Comment (Saransh-Sharma) [inline app/src/androidTest/java/saveServiceTests/StartStopService.java:40]: we could make 3 helper methods 
- start ping service
- start save service
- start both

**Review by Saransh-Sharma — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 8 above)
Comment (sunny-goel) [reply .gitignore]: Done

File: app/src/androidTest/java/saveServiceTests/StartStopService.java (hunk 9 above)
Comment (sunny-goel) [reply app/src/androidTest/java/saveServiceTests/StartStopService.java:40]: Added ServiceHelper.java class which contains all helper methods for start / stop services.

File: .gitignore (hunk 8 above)
Comment (sunny-goel) [reply .gitignore]: Done

File: app/src/androidTest/java/saveServiceTests/StartStopService.java (hunk 9 above)
Comment (sunny-goel) [reply app/src/androidTest/java/saveServiceTests/StartStopService.java:40]: Each and every method of Service class has been tried and tested

**Pull request merged**